### ISSUE TITLE
Back pressure?

### ISSUE THREAD

**Issue by pawamoy** [tiangolo/fastapi#857]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

So, I stumbled upon this article: https://lucumr.pocoo.org/2020/1/1/async-pressure/, posted on Twitter a few days ago by Kenneth Reitz if I recall correctly.

The article talks about back pressure and control flow. I'll let you read it because I'm not knowledgeable enough to summarize it.

I'd like to know if FastAPI is impacted by what is described in this article (possible overloading of the system due to lack of back pressure because async), and if thoughts were put into its design to prevent, or help prevent it.

I'm planning to use FastAPI at work, so it would be nice to have the authors and maintainers point of view on the subject 🙂 . Testimony from FastAPI users on the matter would be great as well! Thanks!
 
EDIT: maybe this should be asked in Starlette instead? Let me know if this is the case, I'll reopen there.

**Comment by dmontagu** [comment]
Yes, FastAPI is affected, but you're right that this is issue is probably better directed to starlette -- for low level details like this, FastAPI basically defers 100% to starlette, so in particular, it should be the case that you'll have backpressure issues in FastAPI if and only if you would in starlette. Given how low level the issue is, it may even be a better issue for uvicorn than for starlette. But I would start by asking in starlette.

For what it's worth, the author of starlette has been aware of this issue for [going on two years](https://news.ycombinator.com/item?id=16907689) so I don't think there is any reason to panic.

While it would be *better* if there was a straightforward way to handle this in the application layer, even if you have very heavy load, in practice you can usually prevent backpressure issues by just using a rate-limiting load balancer and a good auto-scaling policy. This won't handle all cases, and won't save you from particularly bad design choices, but for 99% of deployed python applications this would already be overkill.

I would be curious to hear responses to this issue in the starlette repo though, or (especially) from anyone that has had to deal with backpressure-related issues in python servers.

**Comment by dmontagu** [comment]
For future reference, @tomchristie posted an excellent response in the follow-up starlette issue about how to handle backpressure-related concerns that is fully compatible with FastAPI: https://github.com/encode/starlette/issues/802#issuecomment-574606003

**Comment by tiangolo** [comment]
Thanks @dmontagu ! Very good summary. :bowing_man: :cake:
